Issue/1482 step2 magic link signin changes

anitaa1990 opened the pull request:
Partially fixes #1482. This PR completes the second step of the Jetpack sign in with self credentials as defined in the master thread.

### Changes
- Modified UI in `LoginMagicLinkRequestFragment` and `LoginMagicLinkSentFragment` according to new design.
- Created a new class: `MagicLinkInterceptFragment` that handles the magic link login process. This was previously handled by the `MainActivity`. This logic has been migrated to the new class. 

### Behaviour
(LEFT: Before . RIGHT: After)  
<img width="300" src="https://user-images.githubusercontent.com/22608780/67493103-01aa2a80-f695-11e9-84f1-6b944cbad3c9.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/67492935-cd366e80-f694-11e9-9108-3077c4b56e99.gif">

### Notes:
- The are TODO's added to the PR to add events to the button clicks. This will take place in another PR.
- Note that this PR is to be merged to a feature branch.
- I've kept the option to enter the password as well since the user might not always have the email connected to their account on their phone. This is not part of the design flow though. Let me know if there are any objections and we can certainly hide it.

### Testing:
- Verify that the current flow with magic link is working correctly.
- Verify the current flow to enter password instead of the magic flow link, is working correctly.


Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

anitaa1990 committed:
Design changes for request magic link screen

anitaa1990 committed:
Changed text of magic link sent screen

anitaa1990 committed:
Added new fragment, presenter, module, contract for opening/handling magic link

anitaa1990 committed:
Added logic to open MagicLinkInterceptFragment when magic link is opened

anitaa1990 committed:
Added release notes

anitaa1990 committed:
Merge branch 'feature/sign-in-with-self-hosted-credentials' into issue/1482-step2-magic-link-signin-changes

anitaa1990 committed:
Removed translated string since the base string has been updated

anitaa1990 committed:
Merge branch 'issue/1482-step2-magic-link-signin-changes' of https://github.com/woocommerce/woocommerce-android into issue/1482-step2-magic-link-signin-changes

AmandaRiu commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -496,7 +496,7 @@
     <string name="verification_code">Verification code</string>
     <string name="invalid_verification_code">Invalid verification code</string>
-    <string name="send_link">Send link</string>
+    <string name="send_link">Send verification email</string>
```
🎗  IIRC we'll need to let Platform 9 know there were changes to the _text_ of some of the strings so they can make sure they are re-translated. I believe only new strings are automatically translated.

anitaa1990 replied:
Since this label will be displayed only for self hosted login, I've reverted this string and added a new string label in a5321d2

AmandaRiu commented on libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginMagicLinkRequestFragment.java:
```diff
@@ -143,28 +146,47 @@ public void onClick(View v) {
+
+            TextView labelTextView = view.findViewById(R.id.label);
+            labelTextView.setText(Html.fromHtml(String.format(getResources().getString(
+                    R.string.login_magic_links_label), "<b>", mEmail, "</b>")));
```
There is a more direct approach that could be used for styling this string since the strings resources file does accept basic HTML styling:

```
<string name=“login_magic_links_label”>Almost there! We just need to verify your Jetpack connected email address &lt;b>%1$s&lt;/b></string>
```
then 
```
labelTextView.setText(Html.fromHtml(String.format(getResources().getString(
        R.string.login_magic_links_label), mEmail)));
```

Also, the IDE is really complaining about this line of code and I believe it's because the `login_magic_links_label` string defined in `/lib/login` does not contain any variables - therefore it doesn't need to be formatted. Since we're the only ones using this modified string inside this if statement, maybe it makes sense to have a different string resource defined altogether?

anitaa1990 replied:
Makes sense! Modified in f89abce and added the translated strings for this label again

AmandaRiu commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -517,9 +517,10 @@
     <string name="password_incorrect">It looks like this password is incorrect. Please double check your information and try again.</string>
-    <string name="login_magic_links_label">We\'ll email you a magic link that\'ll log you in instantly, no password needed. Hunt and peck no more!</string>
-    <string name="login_magic_links_sent_label">Your magic link is on its way! Check your email on this device and tap the link in the email you received from WordPress.com.</string>
+    <string name="login_magic_links_label">Almost there! We just need to verify\nyour Jetpack connected email address\n%1$s %2$s %3$s</string>
```
I recommend removing the `\n` characters in the string. It’s ignored in my testing, but if it’s ever not ignored it could have some strange effects on devices with smaller screens, lower res, or have their display/font settings tweaked to make everything bigger.

anitaa1990 replied:
👍 Fixed in f89abce

AmandaRiu reviewed (changes requested):
@anitaa1990 Nice work on all of this! I've finished my first review and have left some comments in the code, plus the following:

## Design Comments

Changes to the magic link view for the self-hosted flow has also changed the magic link view for the normal flow. Was this intentional? If I don’t use my self-hosted auth credentials and just do the normal flow of logging in with my WPcom creds, I see this:

![Screenshot_1571959901](https://user-images.githubusercontent.com/5810477/67533560-4b881600-f67f-11e9-8ace-44782072017b.png)

...which feels wrong in this context since we're _logging in_ and not just verifying an email address. cc: @kyleaparker  for clarification.

## Crash

Error while trying to logout:

``` kotlin
AndroidRuntime  D  Shutting down VM
 E  FATAL EXCEPTION: main
 E  Process: com.woocommerce.android.dev, PID: 10933
 E  org.greenrobot.eventbus.EventBusException: Invoking subscriber failed
 E      at org.greenrobot.eventbus.EventBus.handleSubscriberException(EventBus.java:527)
 E      at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:509)
 E      at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:501)
 E      at org.greenrobot.eventbus.HandlerPoster.handleMessage(HandlerPoster.java:67)
 E      at android.os.Handler.dispatchMessage(Handler.java:106)
 E      at android.os.Looper.loop(Looper.java:193)
 E      at android.app.ActivityThread.main(ActivityThread.java:6669)
 E      at java.lang.reflect.Method.invoke(Native Method)
 E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
 E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
 E  Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.content.Context.getPackageName()' on a null object reference
 E      at android.content.ComponentName.<init>(ComponentName.java:130)
 E      at android.content.Intent.<init>(Intent.java:6082)
 E      at com.woocommerce.android.ui.login.MagicLinkInterceptFragment.showLoginScreen(MagicLinkInterceptFragment.kt:102)
 E      at com.woocommerce.android.ui.login.MagicLinkInterceptPresenter.onAuthenticationChanged(MagicLinkInterceptPresenter.kt:60)
 E      at java.lang.reflect.Method.invoke(Native Method)
 E      at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:507)
 E      ... 8 more
```

To reproduce:
	1. Login with magic link (doesn’t matter which flow)
	2. After successful login, log out. 
	3. Crash.

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptContract.kt:
```diff
@@ -0,0 +1,19 @@
+import com.woocommerce.android.ui.base.BasePresenter
+import com.woocommerce.android.ui.base.BaseView
+
+interface MagicLinkInterceptContract {
```
💯  I thought I probably shouldn't complicate an already complicated feature. But since you mentioned it, I thought we are just migrating the logic to a separate class and I haven't actually played around with MVVM + coroutines yet, so this _would_ be a perfect candidate for MVVM. 

I'm currently working on this now and will update the PR once done. Still a noob 😄

anitaa1990 replied:
I was such a noob and spent most of Friday and a couple of hours on sunday to make this change. But pretty excited about it 😄 Done in 44ba65a

AmandaRiu replied:
Nice!! Yeah that first conversion is super scary, but it's also really exciting and rewarding once it finally "clicks".

kyleaparker commented:
>...which feels wrong in this context since we're logging in and not just verifying an email address. cc: @kyleaparker for clarification.

Yes the copy should be different in the two different scenarios

anitaa1990 committed:
Migrating to MVVM from MVP

anitaa1990 committed:
Added new string labels for magic link sent label

anitaa1990 committed:
Modified release notes

anitaa1990 committed:
Added option to verify email only if login is for self hosted credentials

anitaa1990 commented on RELEASE-NOTES.txt:
```diff
@@ -1,13 +1,13 @@
 3.0
 -----
- 
+* Provided a new option for users to login via their site credentials
```
Done in 8656163

anitaa1990 commented:
Thanks for the thorough review @AmandaRiu 

> Changes to the magic link view for the self-hosted flow has also changed the magic link view for the normal flow. Was this intentional? 

I've fixed this in a5321d2

> Crash Error while trying to logout:

I haven't been able to reproduce this, for some reason. I tried logging in via the normal flow and the self hosted flow and both times it was successful 🤔 I've made some changes to the flow so perhaps it's fixed? Could I trouble you to let me know if you are able to reproduce this now?

PR ready for another round 👍

anitaa1990 committed:
Merge branch 'feature/sign-in-with-self-hosted-credentials' into issue/1482-step2-magic-link-signin-changes

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptRepository.kt:
```diff
@@ -0,0 +1,207 @@
+                    fetchAccountSettings.await()
+                    fetchSites.await()
+
+                    if (fetchedAccountSettingsAndSites) SUCCESS else NO_ACTION_NEEDED
```
The variable `fetchedAccountSettingsAndSites` gets overwritten by each fetch type above and only contains the result of whatever type changed it last. For example, `fetchAccountSettings()` could fail so `fetchedAccountSettingsAndSites` would be `false`, but then `fetchSites()` could be successful leaving `fetchedAccountSettingsAndSites` value at `true`.  To make this more accurate each fetch type should write to it's own variable and then all three variables evaluated in the end to report back whether this was a SUCCESS or not. 

If the login wasn't successful shouldn't we be sending `RequestResult.ERROR` instead of `RequestResult.NO_ACTION_NEEDED` so the user can be notified?

anitaa1990 replied:
Good point! Fixed in 1a9050c

> If the login wasn't successful shouldn't we be sending RequestResult.ERROR instead of RequestResult.NO_ACTION_NEEDED so the user can be notified?

That's a good point! I didn't handle this scenario because this seems to be the behaviour in the current version of the app 😨. We don't seem to be handling error scenarios when we are not able to fetch sites or account info ([#](https://github.com/woocommerce/woocommerce-android/blob/70aad1c2288bb95093f12a327764e3cb8e584856/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt#L138), [#](https://github.com/woocommerce/woocommerce-android/blob/70aad1c2288bb95093f12a327764e3cb8e584856/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt#L166), [#](https://github.com/woocommerce/woocommerce-android/blob/e108de58d3f767f36df0090434f18daeb7b9adbd/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerPresenter.kt#L108)). 

But just to complete the loop on this flow, I added a retry button so that if an error occurs when fetching account/settings/sites, we would display a snackbar error message and provide an option for users to retry the request. 

<img width="300" src="https://user-images.githubusercontent.com/22608780/67830866-5c62dc80-fb02-11e9-9ad1-a7f225a28cce.gif">

AmandaRiu replied:
Nice! I really like the RETRY option!

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptViewModel.kt:
```diff
@@ -0,0 +1,45 @@
+
+                // Errors can occur if the auth token was not updated to the FluxC cache
+                // or if the user is not logged in
+                ERROR -> _isAuthTokenUpdated.value = false
```
Doesn't look like the user is ever notified if there was an error processing the magic link...should they be? If so it may be good to have another LiveData stream for displaying error messages in the UI so you can notify the user like this:

```
private val _showSnackbarMessage = SingleLiveEvent<Int>()
val showSnackbarMessage: LiveData<Int> = _showSnackbarMessage
```

anitaa1990 replied:
Since this was the current flow for magic links, i.e. if we are unable to update the token to Fluxc and if the user is not logged in, we are redirecting them to the login screen without any error messages. But just to complete the magic link flow, I added a snackbar error message as you suggested before redirecting the user to the login screen. Done in 1a9050c

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptFragment.kt:
```diff
@@ -0,0 +1,131 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        view.findViewById<Button>(org.wordpress.android.login.R.id.login_open_email_client).isEnabled = false
```
np: I don't believe you need the full path to these resources...they should all be available to the `com.woocommerce.android` package once a build is completed.

anitaa1990 replied:
🤦‍♀ Fixed in 1a9050c

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptFragment.kt:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    private fun hideProgressDialog() {
+        progressDialog?.apply { if (isShowing) { cancel() } }
```
To prevent memory leaks this should also set `progressDialog = null`.

anitaa1990 replied:
👍 Done in 1a9050c

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptRepository.kt:
```diff
@@ -0,0 +1,207 @@
+    @Subscribe(threadMode = MAIN)
+    fun onAuthenticationChanged(event: OnAuthenticationChanged) {
+        if (event.isError) {
+            continuationUpdateToken?.resume(false)
```
np: Should we be logging these errors?

anitaa1990 replied:
Yup, we should. Events for all the error/success scenarios, the new views/clicks will be added in a separate PR. I meant to mark a TODO here but forgot. Added a TODO in 1a9050c

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptRepository.kt:
```diff
@@ -0,0 +1,207 @@
+        when {
+            event.causeOfChange == AccountAction.FETCH_ACCOUNT -> {
+                // The user's account info has been fetched and stored - next, fetch the user's settings
+                continuationFetchAccount?.resume(!event.isError)
```
np: Should we be logging these errors?

anitaa1990 replied:
Yup, we should. Events for all the error/success scenarios, the new views/clicks will be added in a separate PR. I meant to mark a TODO here but forgot. Added a TODO in 1a9050c

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptRepository.kt:
```diff
@@ -0,0 +1,207 @@
+            }
+            event.causeOfChange == AccountAction.FETCH_SETTINGS -> {
+                // The user's account settings have also been fetched and stored - now we can fetch the user's sites
+                continuationFetchAccountSettings?.resume(!event.isError)
```
np: Should we be logging these errors?

anitaa1990 replied:
Same as below comment.

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptRepository.kt:
```diff
@@ -0,0 +1,207 @@
+    @Subscribe(threadMode = MAIN)
+    fun onSiteChanged(event: OnSiteChanged) {
+        if (event.isError) {
+            continuationFetchSites?.resume(false)
```
np: Should we be logging these errors?

anitaa1990 replied:
Yup, we should. Events for all the error/success scenarios, the new views/clicks will be added in a separate PR. I meant to mark a TODO here but forgot. Added a TODO in 1a9050c

AmandaRiu reviewed (changes requested):
@anitaa1990 Looking great! The conversion to MVVM is so nice! I've only left a few code comments - once addressed this should be ready to merge 👍 

Oh and Im no longer able to reproduce that crash either. Maybe the switch to MVVM fixed it.

anitaa1990 committed:
Added logic to handle error scenarios when fetching account and site info

anitaa1990 commented:
Thanks for the review @AmandaRiu. 

> Just curious why you moved these out of the ActivityBindingModule.

Since we need the `ViewModelModule` and `ThreadModule` across multiple activities now (`MainActivity` and `LoginActivity`) and we might in the future use the same modules if we migrate from MVP to MVVM for the site picker or settings screen, I thought it made more sense to move these two to `AppComponent`

I've fixed the code changes requested and PR is ready for another round 👍

AmandaRiu reviewed (approved):
@anitaa1990 

> Since we need the ViewModelModule and ThreadModule across multiple activities now (MainActivity and LoginActivity) and we might in the future use the same modules if we migrate from MVP to MVVM for the site picker or settings screen, I thought it made more sense to move these two to AppComponent

Makes perfect sense to me! Changes look great...beautiful work! :shipit:
